docs: rewrite readme (#8524)
ibis-project/ibis: README.md

# Ibis

> [![Documentation status](https://img.shields.io/badge/docs-docs.ibis--project.org-blue.svg)](http://ibis-project.org)

Reviewer: Any particular reason to change the case? These don't show up anywhere in the code and are really only for documenting what the link is. You could just as well remove the `Documentation Status` text and keep the inner `[]` empty.

Author: no it just bothered me

[![Project chat](https://img.shields.io/badge/zulip-join_chat-purple.svg?logo=zulip)](https://ibis-project.zulipchat.com)
[![Anaconda badge](https://anaconda.org/conda-forge/ibis-framework/badges/version.svg)](https://anaconda.org/conda-forge/ibis-framework)
[![PyPI](https://img.shields.io/pypi/v/ibis-framework.svg)](https://pypi.org/project/ibis-framework)
[![Build status](https://github.com/ibis-project/ibis/actions/workflows/ibis-main.yml/badge.svg)](https://github.com/ibis-project/ibis/actions/workflows/ibis-main.yml?query=branch%3Amain)
[![Build status](https://github.com/ibis-project/ibis/actions/workflows/ibis-backends.yml/badge.svg)](https://github.com/ibis-project/ibis/actions/workflows/ibis-backends.yml?query=branch%3Amain)
[![Codecov branch](https://img.shields.io/codecov/c/github/ibis-project/ibis/main.svg)](https://codecov.io/gh/ibis-project/ibis)

## What is Ibis?

Ibis is the portable Python dataframe library:

> - Fast local dataframes with the default DuckDB backend

Author: maybe? I'm open to changing it, but I think this serves a few purposes:

- DuckDB hype: DuckDB is awesome, people love it, and Ibis provides a Pythonic dataframe on top by default
- explain the speed -- instead of wondering how Ibis can be so fast (did it implement its own engine?) a user can realize it's from DuckDB
- hint at portability/decoupling from API from engine -- if a user grasps that the Ibis API compiles down to DuckDB SQL, they can hopefully grasp how other backends work

> - Lazy (by default) or interactive execution

Reviewer: I don't like the "or" here, it reads like two aspects of ibis that are somehow mutually exclusive and also related to each other, neither of which is true.

Laziness is a abstract property of computation, and interactive mode is an Ibis-specific feature that is unrelated to laziness.

> - [Mix Python dataframe and SQL code](#python--sql-better-together)

Reviewer: How about "Compose table operations with inline SQL" or "Mix and match inline SQL code with table operations"

> - [Mix Python dataframe and SQL code](#python--sql-better-together)

Author: going with changing "Mix" to "Compose" -- I'm worred "table operations" and "inline" could be confusing to a casual reader

> - [Mix Python dataframe and SQL code](#python--sql-better-together)

Reviewer: Not sure whether we should make decisions about anything based on "casual" readers, but we can leave it at that.

- Use the same dataframe API for [20+ backends](#backends)

> - Iterate locally and deploy remotely by [changing a single line of code](#portability)

Reviewer 2: should we add an example to demonstrate this? e.g. 
1. start with local experimentation
2. change 1 line of code to deploy remotely

Author: yes, though this example would necessarily not work without the user having that remote backend setup and getting their connection (credentials and whatnot) right

we could show DuckDB -> MotherDuck as a simple example

we could also just show like DuckDB -> Snowflake/BigQuery and have a warning about "you need to setup your own remote backend/credentials"

Reviewer: I don't think a code example of this belongs in the readme. We can simply make the claim and then link to a complete example (like [my bigquery + duckdb arrays blogpost](https://ibis-project.org/posts/backend-agnostic-arrays/)).

Author: added a link to it below

> See the documentation on ["Why Ibis?"](https://ibis-project.org/why) to learn more.

Reviewer 2: should we add a graph in readme to explain what ibis is conceptually?

Author: yes but I don't have a good one

there is a nice one here: https://voltrondata.com/resources/open-source-standards

we'd need to de-VoDa-fy it. perhaps we can engage the VoDa design team after GTC? graphs/diagrams would also be good in general for the website

I suggest we don't block this update on having a graph though (there isn't one currently)

Reviewer: If anything, that Sankey diagram seems like a reasonable start. Agree that this shouldn't block.

Author: added Sankey

## Getting started

> You can `pip install` Ibis with a backend and example data:

Author: generally no, and generally you wouldn't want to install `examples`. this installs `gcsfs` and `pins`, the latter being a library Ibis uses for examples which are then read from a public GCS bucket

however for the code below -- `t = ibis.examples.penguins.fetch()` -- this is needed. we could throw a CSV/Parquet file somewhere else for this

> You can `pip install` Ibis with a backend and example data:

Reviewer: > i'd decouple what is needed for ibis and what's just for demonstration purposes

It's already decoupled, that is the purpose of having a separate `examples` [extra](https://setuptools.pypa.io/en/latest/userguide/dependency_management.html#optional-dependencies).

```bash
pip install 'ibis-framework[duckdb,examples]'
```

> > [!TIP]

Reviewer: Neat, didn't know about this syntax!

Author: I learned about it recently!

└───────────┴───────────┴───────┘
>>> t2 = t1.order_by("count")
>>> t2
┏━━━━━━━━━━━┳━━━━━━━━━━━┳━━━━━━━┓
┃ species   ┃ island    ┃ count ┃
┡━━━━━━━━━━━╇━━━━━━━━━━━╇━━━━━━━┩
│ string    │ string    │ int64 │
├───────────┼───────────┼───────┤
│ Adelie    │ Biscoe    │    44 │
│ Adelie    │ Torgersen │    52 │
│ Adelie    │ Dream     │    56 │
│ Chinstrap │ Dream     │    68 │
│ Gentoo    │ Biscoe    │   124 │
└───────────┴───────────┴───────┘
```

This allows you to combine the flexibility of Python with the scale and performance of modern SQL.

## Backends

Ibis supports 20+ backends:

- [Apache Arrow DataFusion](https://ibis-project.org/backends/datafusion/)
- [Apache Druid](https://ibis-project.org/backends/druid/)
- [Apache Flink](https://ibis-project.org/backends/flink)
- [Apache Impala](https://ibis-project.org/backends/impala/)
- [Apache PySpark](https://ibis-project.org/backends/pyspark/)
- [BigQuery](https://ibis-project.org/backends/bigquery/)
- [ClickHouse](https://ibis-project.org/backends/clickhouse/)
- [Dask](https://ibis-project.org/backends/dask/)
- [DuckDB](https://ibis-project.org/backends/duckdb/)
- [Exasol](https://ibis-project.org/backends/exasol)

> - [HeavyAI](https://github.com/heavyai/ibis-heavyai)

Reviewer: Can we remove this? HeavyAI 100% **will not** work with any recent version of Ibis.

Author: yep!

- [MySQL](https://ibis-project.org/backends/mysql/)
- [Oracle](https://ibis-project.org/backends/oracle/)
- [Pandas](https://ibis-project.org/backends/pandas/)
- [Polars](https://ibis-project.org/backends/polars/)
- [PostgreSQL](https://ibis-project.org/backends/postgresql/)
- [SQL Server](https://ibis-project.org/backends/mssql/)
- [SQLite](https://ibis-project.org/backends/sqlite/)
- [Snowflake](https://ibis-project.org/backends/snowflake)
- [Trino](https://ibis-project.org/backends/trino/)

## Portability

You can set the backend Ibis uses:

```python
>>> ibis.set_backend("duckdb")
>>> ibis.set_backend("polars")
>>> ibis.set_backend("datafusion")
```

> More conventionally, you'll create a connection object:

Reviewer: ```suggestion
Typically, you'll create a connection object:
```

```python
>>> con = ibis.duckdb.connect()
>>> con = ibis.polars.connect()
>>> con = ibis.datafusion.connect()
```

And work with tables in that backend:

```python
>>> t = con.table("penguins")
```

> Or, for most backends, read from common file formats like CSV or Parquet:

Reviewer: ```suggestion
You can also read from file formats like CSV and Apache Parquet:
```
